server/src/entities/Party.ts

```diff
@@ -0,0 +1,92 @@
+import {
+  BaseEntity,
+  Column,
+  Entity,
+  ManyToOne,
+  OneToMany,
+  PrimaryGeneratedColumn,
+  Unique,
+  UpdateDateColumn,
+} from 'typeorm'
+import { PlayedSong } from './PlayedSong'
+import { Song } from './Song'
+import { VotedSong } from './VotedSong'
+
+@Unique(['name'])
+@Entity()
+export class Party extends BaseEntity {
+  @PrimaryGeneratedColumn()
+  id: number
+
+  @Column()
+  name: string
+
+  @Column({ nullable: true })
+  password: string | null
+
+  @UpdateDateColumn()
+  latestTime: Date
+
+  @ManyToOne(() => Song, { nullable: true })
+  currentSong: Song | null
+
+  @OneToMany(() => VotedSong, votedSong => votedSong.party, { eager: true, cascade: ['remove'] })
+  votedSongs: VotedSong[]
+
+  @OneToMany(() => PlayedSong, playedSong => playedSong.party, { eager: true, cascade: ['remove'] })
+  playedSongs: PlayedSong[]
+
+  constructor(name: string, password?: string) {
+    super()
+
+    this.name = name
+    this.password = password || null
+    this.currentSong = null
+    this.playedSongs = []
+    this.votedSongs = []
+  }
+
+  public async playNextSong() {
+    await this.removeCurrentSong()
+    await this.playHighestVotedSong()
+  }
+
+  public async voteForSong(song: Song) {
+    let votedSong = await VotedSong.findOne({ song: song, party: this })
+
+    if (votedSong) {
+      await votedSong.incrementVote()
+    } else {
+      votedSong = new VotedSong(song, this)
+      await votedSong.save()
+    }
+
+    return votedSong
+  }
+
+  private async removeCurrentSong() {
+    if (this.currentSong) {
+      const newPlayedSong = new PlayedSong(this.currentSong, this, await this.getNextSequenceNumber())
+      this.currentSong = null
+      await Promise.all([newPlayedSong.save(), this.save()])
+    }
+  }
+
+  private async playHighestVotedSong() {
+    const highestVotedSong = await this.getHighestVotedSong()
+
+    if (highestVotedSong) {
+      this.currentSong = highestVotedSong.song
+      await Promise.all([highestVotedSong.remove(), this.save()])
+    }
+  }
+
+  private async getHighestVotedSong() {
```

Comment:
we need to know the order in case of equivalent number of votes so that we can show the correct order in the frontend